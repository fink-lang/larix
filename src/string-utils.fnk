unindent_text = fn text, ind:
  [first, ...rest] = text.split rx/\n/g

  [...lines] = pipe rest:
    map line:
      line.slice ind

  [first, ...lines].join '\n'
