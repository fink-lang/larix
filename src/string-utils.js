import {curr_loc, next_loc} from '@fink/prattler';


export const unindent_text = (text, ind)=> (
  text
    .split(/\n/g)
    .map((line, idx)=> line.slice(idx > 0 ? ind : 0))
    .join('\n')
);


export const strip_block = (strings, ...parts)=> {
  const [, ...lines] = String.raw({raw: strings}, ...parts).split('\n');
  const ind = lines[0].search(/[^ ]/);
  const str = lines.map((lne)=> lne.slice(ind)).join('\n');
  return str;
};

