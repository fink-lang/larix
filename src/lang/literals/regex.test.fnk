{describe, it, expect, to_match_snapshot, to_throw} = import '@fink/jest'

{parse_expr} = import '../../'


describe:: 'regex rx/.../', fn:

  it:: 'parses empty: rx//', fn:
    expect::
      parse_expr:: `rx//`
      to_match_snapshot


  it:: 'parses multiline', fn:
    expect::
      parse_expr:: `
        rx/
          foo.+bar
        /
      `
      to_match_snapshot


  it:: 'parses flags: rx//gimsuy', fn:
    expect::
      parse_expr:: `rx//gimsuy`
      to_match_snapshot


  it:: 'parses regex: rx/.+foo/', fn:
    expect::
      parse_expr:: `rx/.+foo/gimsuy`
      to_match_snapshot


  it:: 'parses escaped chars: rx/\\n\\t/', fn:
    expect::
      parse_expr:: `rx/\\n \\\\ \\/ \\\\/`
      to_match_snapshot


describe:: 'regex - parsing failures', fn:

  it:: 'throws when missing end', fn:
    expect::
      fn: parse_expr:: `rx/foo.bar`, 'test.fnk'

      to_throw:: `
        test.fnk:1:10
        1| rx/foo.bar
                     ^

        Unexpected end of code.
      `


  it:: 'throws unssupported flags', fn:
    expect::
      fn: parse_expr:: `rx/foo.bar/n`, 'test.fnk'

      to_throw:: `
        test.fnk:1:11
        1| rx/foo.bar/n
                      ^

        Cannot use 'n' as an infix operator.
      `
