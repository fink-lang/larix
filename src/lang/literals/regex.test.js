import {parse_expr} from '../../';
import {strip_block} from '../../string-utils';


describe('regex rx/.../', ()=> {

  it('parses empty: rx//', ()=> {
    expect(
      parse_expr(`rx//`)
    ).toMatchSnapshot();
  });


  it('parses flags: rx//gimsuy', ()=> {
    expect(
      parse_expr(`rx//gimsuy`)
    ).toMatchSnapshot();
  });


  it('parses regex: rx/.+foo/', ()=> {
    expect(
      parse_expr(`rx/.+foo/gimsuy`)
    ).toMatchSnapshot();
  });


  // TODO: fix escapeing \/
  it.skip('parses escaped chars: rx/\\n\\t/', ()=> {
    expect(
      parse_expr(String.raw`rx/\n \\ \/ /`)
    ).toMatchSnapshot();
  });
});


describe('regex - parsing failures', ()=> {

  it('throws when missing end', ()=> {
    expect(
      ()=> parse_expr(`rx/foo.bar`)
    ).toThrow(strip_block`
      Unexpected end of code:
      1| rx/foo.bar
                ^`
    );
  });


  it('throws unssupported flags', ()=> {
    expect(
      ()=> parse_expr(`rx/foo.bar/n`)
    ).toThrow(strip_block`
      Unexpected end of code:
      1| rx/foo.bar/n
                    ^`
    );
  });
});

