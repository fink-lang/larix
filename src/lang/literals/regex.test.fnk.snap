// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`regex rx/.../ parses empty: rx// 1`] = `
regex (1:0-1:4)
  //
`;

exports[`regex rx/.../ parses escaped chars: rx/\\n\\t/ 1`] = `
regex (1:0-1:15)
  /\\n \\\\ \\/ \\\\/
`;

exports[`regex rx/.../ parses flags: rx//gimsuy 1`] = `
regex (1:0-1:10)
  //gimsuy
`;

exports[`regex rx/.../ parses regex: rx/.+foo/ 1`] = `
regex (1:0-1:15)
  /.+foo/gimsuy
`;
